leanprover/lean4
# Auto-compilation & -setup of dependencies on file worker startup

This PR extends the file worker to call `leanpkg print-path <imports>` on startup (if there is a `leanpkg.toml` in the workspace root), which will configure the current package (compiling all dependencies), compile the specified (local) modules, and print the corresponding `LEAN_PATH` to stdout, where it is picked up by the file worker. While running, the latest stderr line of leanpkg is reported as an "information" message.

Note that this won't work when editing Lean's own `src/` because there is no `leanpkg` binary in stage 0.

There are two more minor server changes included:
* A fatal file worker exception after receiving `didOpen` (i.e. when we know the path of the current file) is reported as a diagnostic, which is great for displaying the result of a failed leanpkg invocation
* Only `didChange` requests restart a crashed file worker now; without this change, the `documentSymbol` request threw my Emacs into a restart loop

Feel free to give this a try, perhaps even on Windows if someone feels courageous!

**How to test:**
* open `tests/leanpkg/b/` as a workspace folder
* opening `B.lean` in there should automatically compile & find `B/Foo.lean` as well as the dependency `../a`
* try editing `B/Foo.lean`, e.g. introducing the commented delay or an error; restarting the worker for `B.lean` should make the change visible there

**TODO:**
* [x] Analogous feature using Nix when there is a `flake.nix` in the root instead
